File: problems/sum/services/sum_and_product/python/simple/sum_and_product_server.py
#!/usr/bin/env python3

from os import environ
from sys import stderr, exit
from random import randrange

ENV['numbers'] = environ["TAL_numbers"]
ENV['num_questions'] = int(environ["TAL_num_questions"])

print(f"# I will serve: problem=sum, service=sum_and_product, numbers={ENV['numbers']}, num_questions={ENV['num_questions']}.")

gen_new_pair = True    
for _ in range(ENV['num_questions']):
    if gen_new_pair:
        if ENV['numbers'] == "onedigit":
            x = randrange(10)
            y = randrange(10)
        elif ENV['numbers'] == "twodigits":
            x = randrange(100)
            y = randrange(100)
        else:
            x = randrange(2**32)
            y = randrange(2**32)
    print(f"? {x+y} {x*y}")
    spoon = input().strip()
    while spoon[0] == '#':
        spoon = input().strip()
    a, b = map(int, spoon.split(" "))
    gen_new_pair = False
    if a+b > x+y:
        print(f"No! indeed, {a}+{b}={a+b} > {x+y}.")
    elif a+b < x+y:    
        print(f"No! indeed, {a}+{b}={a+b} < {x+y}.")
    elif a*b > x*y:    
        print(f"No! indeed, {a}*{b}={a*b} > {x*y}.")
    elif a*b < x*y:    
        print(f"No! indeed, {a}*{b}={a*b} < {x*y}.")
    else:
        assert (a + b == x+y) and (a * b == x*y)
        print(f"Ok! indeed, {a}+{b}={x+y} and {a}*{b}={x*y}.")
        gen_new_pair = True

exit(0)

File: example_problems/pure_java/sum/services/sum_and_product/advanced/sum_and_product_server.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Scanner;

public class sum_and_product_server {
    public static void main(String[] args) {
        TALinputs TALinput = new TALinputs();
        String problem = "sum";
        String service= "sum_and_product";
        Map<String,String>args_list=new HashMap<>();
        args_list.put("num_questions","int");
        args_list.put("numbers","String");
        args_list.put("lang","String");
        args_list.put("META_TTY","String");
        System.out.println(args.length);
        Env env= new Env(problem,service,args_list);
        TALcolors TAc = new TALcolors();
        Lang lang = new Lang(env,TAc,service,true);
        TAc.print(lang.opening_msg,"green",null,null);
        System.out.println("");
        int x = 0;
        int y = 0;
        Random randrange = new Random();
        boolean gen_new_pair = true;

        for(int i=0;i <(int)env.ENV("NUM_QUESTIONS");i++) {
            if (env.ENV("NUMBERS").equals("onedigit")) {
                x = randrange.nextInt(10);
                y = randrange.nextInt(10);
            } else if (env.ENV("NUMBERS").equals("twodigits")) {
                x = randrange.nextInt(100);
                y = randrange.nextInt(100);
            } else {
                x = randrange.nextInt(2 * 32);
                y = randrange.nextInt(2 * 32);
            }
            int sum = x + y;
            int prod = x * y;
            List<String> carFormat = new ArrayList<>();
            carFormat.add("bold");
            TAc.print(sum + " " + prod, "yellow", carFormat, null);

            List<Object> vars = new ArrayList<>();
            vars = TALinput.TALinput("int", 2, TAc);
            int a = (int) vars.get(0);
            int b = (int) vars.get(1);
            vars.add(x);
            vars.add(y);
            if (a + b > x + y) {
                TAc.NO();
                TAc.print(lang.render_feedback("over-sum", "No! indeed,%d+%d=%d>%d.", vars), "yellow", null, null);
            } else if (a + b < x + y) {
                TAc.NO();
                TAc.print(lang.render_feedback("under-sum", "No! indeed,%d+%d=%d>%d.", vars), "yellow", null, null);
            } else if (a * b > x * y) {
                TAc.NO();
                TAc.print(lang.render_feedback("over-product", "No! indeed,%d+%d=%d>%d.", vars), "yellow", null, null);
            } else if (a * b < x * y) {
                TAc.NO();
                TAc.print(lang.render_feedback("under-product", "No! indeed,%d+%d=%d>%d.", vars), "yellow", null, null);
            } else {
                TAc.OK();
                TAc.print(lang.render_feedback("ok", "No! indeed,%d+%d=%d>%d.", vars), "yellow", null, null);
                gen_new_pair = false;
            }
        }
        TAc.Finished();
        System.exit(0);
    }
}
